### PR TITLE
Refactor of Image Matching Algorithm

### PR DESCRIPTION
**Pull request by clarakosi**
Changes:
* Adds spark udf
* Modifies schema for top_candidates column to now view null
image suggestions as an empty array
* Saves output as parquet in hdfs

Does not enable cluster mode in spark because it does not appear to be possible with jupyter notebooks

### CODE REVIEW

File: etl/schema.py [hunk 1]
```diff
@@ -1,21 +1,18 @@
-from pyspark.sql.types import StructType, StringType
+from pyspark.sql.types import StructType, StringType, ArrayType, MapType, LongType
 
 
-class CsvDataset:
+class ParquetDataset:
     schema = (
         StructType()
-        .add("pandas_idx", StringType(), True)
```
Comment (gmodena) [inline etl/schema.py:7]: Nice that we got rid of this!

FYI: these changes to spark & hql code will need to be incorporated into https://gitlab.wikimedia.org/gmodena/platform-airflow-dags/-/tree/multi-project-dags-repo/image-matching

File: etl/transform.py [hunk 2]
```diff
@@ -62,11 +62,11 @@ def __init__(self, dataFrame: DataFrame):
 
     def transform(self) -> DataFrame:
         with_recommendations = (
-            self.dataFrame.where(~F.col("top_candidates").isNull())
+            self.dataFrame.where(F.size(F.col("top_candidates")) > 0)
```
Comment (gmodena) [inline etl/transform.py:65]: Here you removed a null check, but `top_candidates` schema says the column could be nullable. 
I think your approach here is the correct one, but could we make the schema consistent?

**Review by gmodena — APPROVED** [review]
LGTM! 

I have a small comment re schema, that can be adressed once we move those bits of code out of this repo.

The notebook & udf refactoring look great! There's a couple of things that will need changes once we convert it to a script, but we can discuss those separarely.

File: etl/transform.py (hunk 2 above)
Comment (clarakosi) [reply etl/transform.py:65]: Sure!

File: etl/schema.py (hunk 1 above)
Comment (clarakosi) [reply etl/schema.py:7]: I'm actually not going to merge this PR into this repo but keep this one consistent with version one of the algorithm. I'll make PR on your GitLab repo

**Pull request closed without merging**